### PR TITLE
add icon classes and replace src-matching by class-based matching

### PR DESCRIPTION
**Pull request by clouless**
### Improvement

This is a suggestion to add `class="icon-fav-active"` and `class="icon-fav-inactive"` to the `<img` and replace the JS functionality that matches for `src=star.png` by a classname matching as it is done with all other icons in jenkins.

I need this for my theme development and I think this would be an improvement

**UPDATE**

 * I compiled it and tested in Jenkins v 2.32 see preview below
 * It works as expected
 * (And I fixed the built in `Favorites.java` which already had an error I did not cause)

**UPDATE 2**

 * The icon size classess are also set with `icon-md`, `icon-lg` a.s.o
 * Event is dispatched when fav-icon changes from active to inactive and vice versa
   * `window.dispatchEvent(new CustomEvent('favorite-plugin-icon-change', { transition: "inactive-to-active" }));`
   * With this code you can subscribe to the event 

```
window.addEventListener("favorite-plugin-icon-change", function(){
    console.log('EVENT RECORDED');
});
```

### Preview

favorite-with-class

### CODE REVIEW

**Comment by i386** [comment]
@clouless thanks for contributing!

I think this change could be cleaner if you replaced the `img` tag with a `div` and had CSS setting the background of the div. That way we can remove all the code that swaps out the `img src=`

File: src/main/java/hudson/plugins/favorite/Favorites.java [hunk 1]
```diff
@@ -136,6 +136,10 @@ public boolean hasNext() {
                     public Item next() {
                         return jenkins.getItemByFullName(iterator.next());
                     }
+
+                    /* Java 7 compatibility fallback */
+                    @Override
+                    public void remove() { throw new UnsupportedOperationException(); }
```
Comment (i386) [inline src/main/java/hudson/plugins/favorite/Favorites.java]: Why is this change necessary?

Comment (clouless) [inline src/main/java/hudson/plugins/favorite/Favorites.java]: The built failed because of that. And after some googling it seems that depending on the JDK that builds the plugin this is needed. When built with JDK 8 it is not necessary.
When building with JDK 7 and lesser it is. So either fix the jenkins or keep this :)

**Comment by clouless** [comment]
@i386 ok nice. I will change it to div with background tomorrow. Since this will involve more testing on all major browsers. I will report back here :)

**Comment by i386** [comment]
@clouless great! I think you can get rid of `getSizeClassFromIconSize` stuff too and just serve the largest asset and let CSS scale it appropriately. Even delete the smaller assets.

**Comment by clouless** [comment]
@i368 Yes only one big asset would be fine. But i would like to keep the `getSizeClassFromIconSize` thing since this gives me the correct size in compliance with other jenkins icons. These classes can then be used to scale the big asset by CSS.

In the classic Jenkins you have the "S", "M" and "L" size switcher below all the jobs and the logic is to set these classnames on icons. I think that is a good convention which only the favorite-plugin does not seem to follow. So mainly I want it to follow the convention (even if the classes would be set and have no further use other than fullfilling the convention)

File: src/main/java/hudson/plugins/favorite/Favorites.java (hunk 1 above)
Comment (i386) [inline src/main/java/hudson/plugins/favorite/Favorites.java]: Can you please submit this change as a different PR?

Comment (i386) [inline src/main/java/hudson/plugins/favorite/Favorites.java]: Great I'll approve it straight away

Comment (clouless) [inline src/main/java/hudson/plugins/favorite/Favorites.java]: done with #7

**Comment by clouless** [comment]
I will need to refactor the plugin even further. I will do:

 * Create `css.jelly` like in https://github.com/jenkinsci/sloccount-plugin/tree/master/src/main/resources/tabview
 * Create `javascript.jelly` with JS and move i18n to *Message_*.properties*
 * Remove *.js files with duplicated code

**Comment by clouless** [comment]
**UPDATE** 

  * I did all that and fixed i18n.
  * I will now merge master to resolve conflicts.

@i386 I added a README.md and a LICENSE.md with MIT License. This should be done for all plugin repos btw. :D

File: LICENSE.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# The MIT License (MIT)
```
Comment (i386) [inline LICENSE.md]: @larrys is this the original licence?

File: LICENSE.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+Copyright (c) `2016` `CloudBees, Inc.`
```
Comment (i386) [inline LICENSE.md]: 🐛  This copyright does not belong to CloudBees

Perhaps
```
Copyright (c) 2010 Larry Shatzer, Jr.
```

File: pom.xml [hunk 4]
```diff
@@ -16,6 +16,12 @@
     <name>Favorite</name>
     <url>http://wiki.jenkins-ci.org/display/JENKINS/Favorite+Plugin</url>
 
+    <licenses>
```
Comment (i386) [inline pom.xml]: Please remove all the license attribution changes, including the LICENSE.md and put in another PR until I can get @larrys to tell us the license.

File: src/main/webapp/favoriteToggle.js [hunk 5]
```diff
@@ -1,12 +1 @@
...
-  if(image.src.match(/star.png$/)) {
-    image.src = image.src.replace("star.png", "star-gold.png");
-    hoverNotification('Favorite added', a.parentNode);
-  } else {
-    image.src = image.src.replace("star-gold.png", "star.png");
-    hoverNotification('Favorite deleted', a.parentNode);
-  }
-  return false;
-}
+/* MOVED: src/main/resources/hudson/plugins/favorite/assets/javascript.jelly */
```
Comment (i386) [inline src/main/webapp/favoriteToggle.js]: Just delete the file

File: src/main/resources/hudson/plugins/favorite/assets/css.jelly [hunk 6]
```diff
@@ -0,0 +1,49 @@
+<?jelly escape-by-default='true'?>
```
Comment (i386) [inline src/main/resources/hudson/plugins/favorite/assets/css.jelly]: The favorite star is not vertically aligned

dashboard jenkins 2016-11-22 08-16-33

File: pom.xml [hunk 7]
```diff
@@ -71,8 +77,18 @@
...
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>junit</artifactId>
+            <version>1.19</version>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
-
+    <properties>
+        <jenkins.version>2.32</jenkins.version>
```
Comment (i386) [inline pom.xml]: Please remove the version as it would make this plugin incompatible with Blue Ocean which needs 2.7.1 or above.

If there is a good reason for doing so (for example, CSS or JavaScript needs Jenkins 2) then set this to 2.7.1.

File: pom.xml [hunk 8]
```diff
@@ -71,8 +77,18 @@
...
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
```
Comment (i386) [inline pom.xml]: This comes from the base POM. Please remove.

File: pom.xml [hunk 9]
```diff
@@ -71,8 +77,18 @@
...
+        <hpi-plugin.version>1.120</hpi-plugin.version>
```
Comment (i386) [inline pom.xml]: Also comes from the parent pom. Please remove.

File: pom.xml [hunk 10]
```diff
@@ -71,8 +77,18 @@
             <version>1.5.1</version>
             <optional>true</optional>
         </dependency>
+        <dependency>
```
Comment (i386) [inline pom.xml]: Surprised that you needed a new JUnit. Why is this needed? We should always use the version in the parent pom

The version the parent POM lists is:
```
     <dependency>
        <groupId>junit</groupId>
        <artifactId>junit</artifactId>
        <version>4.12</version>
      </dependency>
```

File: src/main/java/hudson/plugins/favorite/Favorites.java [hunk 11]
```diff
@@ -136,7 +136,7 @@ public boolean hasNext() {
                     public Item next() {
                         return jenkins.getItemByFullName(iterator.next());
                     }
-                    
+
```
Comment (i386) [inline src/main/java/hudson/plugins/favorite/Favorites.java]: Undo this whitespace change

File: src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly [hunk 12]
```diff
@@ -3,10 +3,11 @@
          xmlns:t="/lib/hudson" xmlns:s="/lib/form">
   <j:if test="${it.isLoggedIn()}">
     <j:set var="sortData" value="${it.getSortData(job.getFullName())}"/>
-    <j:set var="star" value="${it.getStar(job.getFullName())}"/>
+    <j:set var="starClass" value="${it.getStarClassName(job.getFullName())}"/>
+    <j:set var="sizeClass" value="${it.getSizeClassFromIconSize(subIconSize)}"/>
     <td data="${sortData}">
       <a onclick="return toggleFavorite('${job.getFullName()}', '${id.userId}', this)" href="#">
-        <img src="${imagesURL}/${subIconSize}/${star}" border="0" id="fav_${job.getFullName()}"/>
+        <span id="fav_${job.getFullName()}" class="${starClass} ${sizeClass}"></span>
```
Comment (i386) [inline src/main/resources/hudson/plugins/favorite/column/FavoriteColumn/column.jelly]: Nice 💯

File: LICENSE.md (hunk 3 above)
Comment (i386) [inline LICENSE.md]: Lets move LICENSE changes to another PR so I can get feedback from @larrys (it's his IP)

File: pom.xml (hunk 10 above)
Comment (clouless) [inline pom.xml]: I had problems when running `mvn hpi:run` telling me I needed this, but I can remove it.

File: pom.xml (hunk 4 above)
Comment (clouless) [inline pom.xml]: ok I will remove all license stuff and move to separate PR.

File: LICENSE.md (hunk 3 above)
Comment (clouless) [inline LICENSE.md]: ok I will remove all license stuff and move to separate PR.

File: LICENSE.md (hunk 2 above)
Comment (clouless) [inline LICENSE.md]: ok I will remove all license stuff and move to separate PR.

File: pom.xml (hunk 7 above)
Comment (clouless) [inline pom.xml]: I just moved this stuff inside because `mvn:hpi run` had problems otherwise. But I could start `mvn hpi:run` with parameters instead of defining it in the pom.

File: src/main/webapp/favoriteToggle.js (hunk 5 above)
Comment (clouless) [inline src/main/webapp/favoriteToggle.js]: ok

File: src/main/java/hudson/plugins/favorite/Favorites.java (hunk 11 above)
Comment (clouless) [inline src/main/java/hudson/plugins/favorite/Favorites.java]: ok

File: src/main/resources/hudson/plugins/favorite/assets/css.jelly (hunk 6 above)
Comment (clouless) [inline src/main/resources/hudson/plugins/favorite/assets/css.jelly]: fixed. icons align vertical now.

**Comment by clouless** [comment]
@i386 I fixed all requested things. Let's do the LICENSE stuff later when this PR is fully merged.

Sorry for the Time-delay. Being in Germany and you seem to be in Australia makes the communication very asynchronous. I hope that is ok :)

File: src/main/resources/hudson/plugins/favorite/assets/css.jelly (hunk 6 above)
Comment (i386) [inline src/main/resources/hudson/plugins/favorite/assets/css.jelly]: Verified here - thanks!

**Comment by i386** [comment]
I used `mvn hpi:run`

**Comment by i386** [comment]
This LGTM 💯

**Comment by i386** [comment]
Can you please squash all the commits and rebase against master?

**Comment by clouless** [comment]
@i386 ok will do. Do you have any commandline git commands for me? I don't want to mess it up ^^

**Comment by i386** [comment]
I merged this in #9

**Comment by i386** [comment]
Released as favorite 2.0.3 - thanks!

**Comment by clouless** [comment]
Thanks :) And great job with blue ocean btw. I use it a lot and will write a blogpost about it on wednesday :)

**Comment by i386** [comment]
@clouless thank you! Oooh do link me the blog post when its up 👍

**Comment by i386** [comment]
@clouless looks like this changed introduced a bug. Would you mind taking a look? https://issues.jenkins-ci.org/browse/JENKINS-39990

**Comment by clouless** [comment]
@i386 sure. I will provide a new PR with a fix right away.

**Comment by i386** [comment]
@clouless thanks for jumping on this so quick 🎉

**Comment by clouless** [comment]
@i386 no problem :D sorry for breaking it 😼
